testing/file_test/file_test_base.cpp

```diff
@@ -265,151 +266,119 @@ auto FileTestBase::DoArgReplacements(
   return Success();
 }
 
-auto FileTestBase::ProcessTestFile(TestContext& context) -> ErrorOr<Success> {
-  // Original file content, and a cursor for walking through it.
-  llvm::StringRef file_content = context.input_content;
-  llvm::StringRef cursor = file_content;
+// Processes conflict markers, including tracking of whether code is within a
+// conflict marker. Returns true if the line is consumed.
+static auto TryConsumeConflictMarker(llvm::StringRef line,
+                                     llvm::StringRef line_trimmed,
+                                     bool* inside_conflict_marker)
+    -> ErrorOr<bool> {
+  bool is_start = line.starts_with("<<<<<<<");
+  bool is_middle = line.starts_with("=======");
+  bool is_end = line.starts_with(">>>>>>>");
+
+  // When running the test, any conflict marker is an error.
+  if (!absl::GetFlag(FLAGS_autoupdate) && (is_start || is_middle || is_end)) {
+    return ErrorBuilder() << "Conflict marker found:\n" << line;
+  }
+
+  // Autoupdate tracks conflict markers for context, and will discard
+  // conflicting lines when it can autoupdate them.
+  if (*inside_conflict_marker) {
+    if (is_start) {
+      return ErrorBuilder() << "Unexpected conflict marker inside conflict:\n"
+                            << line;
+    }
+    if (is_middle) {
+      return true;
+    }
+    if (is_end) {
+      *inside_conflict_marker = false;
+      return true;
+    }
+
+    // Look for CHECK lines, which can be discarded.
+    if (line_trimmed.starts_with("// CHECK:STDOUT:") ||
+        line_trimmed.starts_with("// CHECK:STDERR:")) {
+      return true;
+    }
```

Comment:
This somewhat duplicates the handling in `TryConsumeCheck`. I think that can be avoided though: if you reorder the call to `TryConsumeCheck` to before `TryConsumeConflictMarker`, then you could just error here if you see *any* line that's not a conflict marker.

Reply:
Including the CHECK lines from both sides of the conflict marker doesn't seem better or worse than including the CHECK lines from neither side to me -- either way, line numbering for the rest of the file will be wrong. And I don't think it matters: we only remove conflict markers in autoupdate mode, in which case the list of old check lines shouldn't be used for anything.

Or am I missing something?

(I don't mean to push hard on this; while I think it'd be a bit cleaner to not repeat these CHECK line checks here, I think it's not a big deal.)

Reply:
OK, looking at the rest of the updated PR, I think this suggestion doesn't work any more, because you want to remove the conflict lines but not the check lines before you handle file splits and build the input file contents.

Reply:
Note, before I think I was also focused on the line_index increment and wanting to have one place to do that. It'd need to be before CHECK is processed but shouldn't be done until after conflict markers decide whether to do it. That could be addressed by conditionality around incrementing the line_index, or doing it in more locations. But I think any approach here ends up with some duplication of logic due to overlap in what's going on.